Allow scope to be passed as array

SiebelsTim opened the pull request:
<!--
Describe the big picture of your changes here to communicate to the maintainers why we should accept this pull request.

This text will be included in the changelog. If applicable, include links to documentation or pieces of code.
If your change includes breaking changes please add a code block documenting the breaking change:

```
BREAKING CHANGES: This patch changes the behavior of configuration item `foo` to do bar. To keep the existing
behavior please do baz.
```
-->

Scopes are currently passed as a scope string, separating scopes by spaces.
Clients can grow to many scopes, resulting in a very long string.

This change allows us to specify scopes using the property scopeArray. That way, we can separate scopes by newlines.
Additionally, this allows us to comment a single scope temporarily or add a comment for a specific scope, e.g. as a reason why that client has this scope granted.

## Checklist

<!--
Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of
them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code.
-->

- [x] I have read the [contributing guidelines](../blob/master/CONTRIBUTING.md) and signed the CLA.
- [ ] I have referenced an issue containing the design document if my change introduces a new feature.
- [x] I have read the [security policy](../security/policy).
- [x] I confirm that this pull request does not address a security vulnerability. 
      If this pull request addresses a security vulnerability, 
      I confirm that I got approval (please contact [security@ory.sh](mailto:security@ory.sh)) from the maintainers to push the changes.
- [x] I have added tests that prove my fix is effective or that my feature works.
- [x] I have added the necessary documentation within the code base (if appropriate). -- Is kubectl explain enough?

Demonsthere commented:
Hi there! 
I am kinda conflicted here 😕. On one hand, it is a pretty simple change, and we can have both scopes and scropeArray in the CR, on the other hand a proper solution imho would be to add a mutationwebhook, which would transform the CR into the new format, wdyt?

SiebelsTim commented:
Hi!
I think the greatest downside is having both fields present in the CR. That is prone to bugs. 

I don't know if we can solve this though. I don't have much experience with mutation webhooks, but from what I see, we would need to have both fields present in the CRD to be able to transmit the payload to the API server. Only if that is given, the webhook is called; otherwise it is denied by the API server directly.

Considering that, introducing mutating webhooks doesn't seem worth the effort to me. The only upside I see is that the resulting resource is clean: It would only contain one of the fields, not both. Am I missing something here?

Demonsthere commented:
After some thought, i think the api is in a early enough version, that we can leave both and  mark the scopes fields as deprecated, removing it after some time :)

SiebelsTim commented:
Sounds good to me. I updated the PR accordingly.

I also modified the behaviour slightly: Both scope properties are merged, rather than overwriting `scope` with `scopeArray`. What do you think?

SiebelsTim commented:
I cannot reproduce the test failures locally. Are they known to be flaky in the CI?

Demonsthere commented:
I think this can be related to go version looking at the logs 🤔 We are pinned to 1.22.0 still

SiebelsTim commented:
I was able to reproduce this once with 1.22.8 (previously using 1.22.2). Might've been a fluke, as I couldn't reproduce this a second time.

For the CI, I simply pushed again to retrigger the CI and it works :shrug:

Demonsthere commented on api/v1alpha1/oauth2client_types.go:
```diff
@@ -145,12 +145,17 @@ type OAuth2ClientSpec struct {
+	// +kubebuilder:validation:Pattern=([a-zA-Z0-9\.\*]+\s?)*
 	//
-	// Scope is a string containing a space-separated list of scope values (as
+	// [DEPRECATED] Scope is a string containing a space-separated list of scope values (as
```
we can use the kubebuilder annotation here
```
// +kubebuilder:deprecatedversion
warningstring
marks this version as deprecated.
```

SiebelsTim replied:
Are you sure this works? I tried that but it had no effect on the generated manifest. I have a feeling this only works on the entire config object, not a single field.

I optimistically changed it anyway.

Demonsthere replied:
Ok i might have misread that 🤔 
```
//+kubebuilder:deprecatedversion:warning="example.com/v1alpha1 CronTab is deprecated" 
```
this might be set on the whole apiversion, and not a single field in the spec, as the openapi schema allows for field deprecation, but kubebuilder may not yet have the tooling for this 😞

Demonsthere reviewed (changes requested):
Please add the deprecated marker, and I think we can merge this :) For the version issue, i will open a separate issue to bump go
